File: CTCI/Ch3StackAndQueue/MinStack.java
import java.util.*;

/**
* This is a stack implementation that returns
* the miniumum value in O(1) time.
*
* Ex:
* TOP: 11, 13, 4, 5, 8, 9 --> min = 4
*/
public class MinStack {
    ArrayList<Integer> stack = new ArrayList<>();
    ArrayList<Integer> minStack = new ArrayList<>();
    int min = Integer.MAX_VALUE;

    public void push(int data) {
        if (data < min) {
            min = data;
            minStack.add(data);
        }
        stack.add(data);
    }

    public int peek() {
        return stack.get(stack.size() - 1);
    }

    public int pop() {
        int poppedVal = stack.get(stack.size() - 1);
        stack.remove(stack.size() - 1);
        if (poppedVal == min) {
            minStack.remove(minStack.size() - 1);
            min = minStack.size() - 1; // Assign the next in line for min value
        }
        return poppedVal;
    }

    public int min() {
        return min;
    }

    public static void main(String[] args) {
        MinStack ms = new MinStack();
        ms.push(10);
        ms.push(4);
        ms.push(19);
        ms.push(1);
        ms.push(11);
        System.out.println(ms.peek());
        System.out.println(ms.min());
    }
}

File: CTCI/Ch3StackAndQueue/Py/min_stack.py
class Stack(object):
    def __init__(self):
        self._min = 100000000000
        self.stack = []
        self.historic_mins = []

    def push(self, item):
        if item not in self.stack:
            self.stack.append(item)
            if item < self._min:
                self.historic_mins.append(item)
                self._min = item

    def pop(self):
        old_min = self.stack.pop()
        if self._min == old_min:
            self.historic_mins.pop()
            self._min = self.historic_mins[0]

    @property
    def min(self):
        return self._min

    @property
    def print_stack(self):
        print(self.stack)


stack = Stack()
stack.push(1)
stack.push(2)
print(stack.historic_mins)
stack.push(3)
print(stack.min)
stack.print_stack
stack.push(-5)
stack.push(6)
print(stack.historic_mins)
print(stack.min)
stack.print_stack
stack.pop()
stack.pop()
print(stack.historic_mins)
print(stack.min)
stack.print_stack
